File: src/sort/first/QuickSort.java
package misc.first;

import java.util.Arrays;

import tool.CommonTool;

/**
 * quick sort
 * 
 * @since 2020-8-5 Wednesday 16:33 - 17:34
 */
public class QuickSort {
    public static void sort(int[] arr) {
        if (arr == null) {
            return;
        }
        helper(arr, 0, arr.length - 1);
    }

    private static void helper(int[] arr, int start, int end) {
        if (start >= end) {
            return;
        }
        int small = start - 1, large = end;
        int i = start, j = end - 1;
        while (i <= j) {
            if (arr[i] > arr[end]) {
                CommonTool.swap(arr, i, j--);
                large--;
            } else if (arr[i] < arr[end]) {
                CommonTool.swap(arr, ++small, i++);
            } else {
                i++;
            }
        }
        CommonTool.swap(arr, large++, end);
        helper(arr, start, small);
        helper(arr, large, end);
    }

    public static void main(String[] args) {
        int[] a1 = new int[] { 5, 4, 2, 1, 51, 2 };
        int[] a2 = new int[] { 5, 2, 4, 1, 2, 1, 5, 1, 1, 1541 };
        int[] a3 = new int[] { 5, 2 };
        sort(a1);
        sort(a2);
        sort(a3);
        System.out.println(Arrays.toString(a1));
        System.out.println(Arrays.toString(a2));
        System.out.println(Arrays.toString(a3));
    }
}

File: src/sort/second/QuickSort.cpp
#include <cstdlib>
#include <iostream>

#include "../Tool.h"

/**
 * quick sort
 *
 * @since 2020-9-24 Thursday
 */
class QuickSort {
   private:
    void swap(int* arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    void helper(int* arr, int start, int end) {
        if (start >= end) {
            return;
        }
        int pivot = std::rand() % (end - start + 1) + start;
        int small = start - 1, big = end, i = start, j = end - 1;
        swap(arr, pivot, end);
        while (i <= j) {
            if (arr[i] > arr[end]) {
                swap(arr, i, j--);
                big--;
            } else if (arr[i] < arr[end]) {
                swap(arr, i++, ++small);
            } else {
                i++;
            }
        }
        swap(arr, big++, end);
        helper(arr, start, small);
        helper(arr, big, end);
    }

   public:
    void sort(int* arr, int len) {
        helper(arr, 0, len - 1);
    }
};

int main(int argc, char const* argv[]) {
    QuickSort s;
    int arr[7] = {9, 6, 1, 1, 10, 123, 6};
    s.sort(arr, 7);
    printArr(arr, 7);

    int worst[7] = {7, 6, 5, 4, 3, 2, 1};
    s.sort(worst, 7);
    printArr(worst, 7);
}
